rfd/0212-jsonpath-interpolation.md

```diff
@@ -0,0 +1,635 @@
+there.
+
+By using a project close to the specification, we can selectively rely on
+official JSONPath documentation, like those linked above, as well as the
+[online sandboxes](https://serdejsonpath.live/).
+
+Note: We must be careful when updating the upstream library, as some of the
+syntax may be subject to change as it adjusts to the RFC and some of the
+unresolved syntax disagreements in the community.
+
+### OIDC Claims
+
+During OIDC login, a user's OIDC claims serve two purposes:
+
+* Setting claims as Teleport user traits, optionally using login rules for custom claims to traits mapping.
+* Determining what Teleport roles to give the user, using the OIDC connector's `claims_to_roles` field to map user traits to roles.
+
+The `jsonpath` function will be supported in login rule trait mapping. As a
+result, any JSON OIDC claims mapped to traits in the claims to traits mapping
+will be available in the traits to roles mapping.
+
+Note: see [this section](#json-traits) for an explanation as to why we decided
+not to set arbitrary JSON values as user traits directly, and instead are
+requiring the use of login rules.
+
+#### Login Rules - Claims to Traits
+
+You can use a login rule to map JSON claims to users traits.
+
+In the example below, a JSON claim is mapped to individual user traits.
+
+```json
+{
+  // groups is a JSON object rather than a string array.
+  "groups": {
+    "roles": ["template"],
+    "logins": ["alice"],
+    "env": ["staging", "dev"],
+  }
+}
+```
+
+```yaml
+kind: login_rule
+version: v1
+metadata:
+  name: my-loginrule
+spec:
+  priority: 0
+  traits_map:
+    roles:
+      # evaluates to ["template"]
+      - jsonpath("$.groups.roles")
+    logins:
+      # evaluates to ["alice"]
+      - jsonpath("$.groups.logins")
+    env:
+      # evaluates to ["staging", "dev"]
+      - jsonpath("$.groups.env")
+```
+
+These traits can then be used in claims to roles mappings, role templates,
+label expressions, etc.
+
+```yaml
+kind: oidc 
+version: v2
+metadata:
+  name: my-idp
+spec:
+  ...
+  claims_to_roles:
+    - claim: "roles"
+      value: "template"
+      roles: ["template"]
+```
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: template
+spec:
+  ...
+  allow:
+    logins: '{{external.logins}}'
+    node_labels_expression: 'contains(external.env, labels["env"])'
+```
+
+### UX
+
+#### User stories
+
+The user stories below will explore example custom OIDC solutions with
+potential Teleport configurations to consume the custom OIDC claims using the
+new `jsonpath` function.
+
+##### Example: IdP with arbitrary JSON claims
+
+Let's say we have a custom IdP which directly supports arbitrary JSON claims to
+be set for users. Below is an example claim object for user `alice`.
+
+```json
+{
+  "groups": {
+    "teleport": {
+      "roles": ["template"],
+      "node": {
+        "logins": "alice",
+        "labels": {
+          "*": "*"
```

Comment:
nit: I think it would have helped me understand if this line was:
`"somekey": "*"`. It was not obvious to me whether the `[?(@ == '*')]` syntax refers to the object key (attribute name) or value (attribute value), but from reading the spec its the value I believe.

Reply:
Actually I think this was just a mistake...I probably realized `labels.*` wouldn't work and overcomplicated it, but it just needs to be `labels['*'].

Reply:
Oh, so that means I had it wrong? The `"*"` refers to the left-hand side (key) and not to the right-hand side (value)?